cpp/test/CMakeLists.txt

```diff
@@ -244,6 +244,8 @@ if(BUILD_TESTS)
     test/neighbors/ann_ivf_flat/test_float_int64_t.cu
     test/neighbors/ann_ivf_flat/test_int8_t_int64_t.cu
     test/neighbors/ann_ivf_flat/test_uint8_t_int64_t.cu
+    test/neighbors/ann_ivf_pq/test_float_int64_t.cu
```

Comment:
Realistically, do we expect to see users using `int64_t` index types when `uint64_t` is already available? Or am I misunderstanding the intention of this index param type? I propose we standardize upon one of them and use them consistently across the APIs and documentation. I'd personally prefer to use `int64_t` over `uint64_t` because that's what I believe faiss has recently switched (back) to using across the board ([ref](https://github.com/facebookresearch/faiss/commit/43d3dcf06184a6351d1fb0e8277c1238c71476b6)). I think we should also consider standardizing on these types across raft in general. I've been considering ways to document and communicate the specialized types to users and it would be great if we could send a consistent message. 

While these additional instantiations may not seem like a lot of additional time on a single architecture, please consider that CI is building for 6 different architectures and `sccache` has been, I fear` sending us a false message as to the amount of time it actually takes to build the codebase end-to-end. An additional instantiation which makes a test take 5 minutes longer to compile for a single architecture could add 20+ minutes to CI. 5 different instantiations which make this assumption could extend CI by over an hour. Right now the challenge is that RAFT is being statically built into both cugraph and cuml's wheels at the moment so these additional compile times extend further than just raft's CI.

Reply:
I don't particularly want test `int64_t` specifically, but I still would like to keep at least one extra test with another index type besides `uint64_t` (at the moment I have two here). Maybe I'd remove this, but keep `uint32_t` version?
TL;DR

I totally agree with our move to stick to one index type, and I also slightly prefer a signed 64 bit type, because it's a little bit safer (e.g. against underflows in reverse loops).

Something I hope but haven't thoroughly tested yet, is that setting a different index type should add a negligible extra compile time compared to the similarity kernel (because the index type is converted to `uint32_t` anyway). Also I surely don't want to add these extra instances to the distance/specializations library, but keep it only in tests (so that it does not affect downstream projects).

At the same time, I appreciate the raft's flexible interface providing value/index template parameters. That allows the downstream projects like faiss to stick to their indexing types and avoid doing extra data conversion passes on every raft invocation and simplify the integration in general. I like that our model of flexible templates + limited specializations allows both alternatives, either fast compile times using raft shared library or flexible instances and compilation on the downstream project side.
The problem we have already encountered in raft in the past, is that we provide flexible templates but test them  with only one template instance. Then, some of the functions don't even compile with another instance, because somewhere in the middle of the code we hardcode something like `rmm::device_uvector<uint64_t> tmp_buffer` instead of `rmm::device_uvector<IdxT>` by accident.

Reply:
Yep, I understand your point here and I'm on board with this. I think we can find a good compromise here- maybe choose an indexing type that we use consistently but have another that can still test. 

As per offline conversation I just wanted to note here that we have been discussing having a set of tests for an expanded set of types that are compiled one nightly to save precious CI time.